### PR TITLE
[rllib] Add async remote workers

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Add a variant of `remote_worker_envs`, `async_remote_worker_envs`, that does dynamic batching as envs complete. It also supports multi-agent.

## Related issue number

Closes https://github.com/ray-project/ray/issues/4235

### CODE REVIEW

**Comment by ericl** [comment]
@bjg2 this unifies the vectorized and multi-agent code paths for remote envs, and adds support for asynchronous execution. Let me know if I missed anything.

**Comment by bjg2** [comment]
> @bjg2 this unifies the vectorized and multi-agent code paths for remote envs, and adds support for asynchronous execution. Let me know if I missed anything.

Looks good! I'm just not sure naming nailed it. Word Vector is omitted and Async is in there even though it might be sync (it is an argument). Maybe RemoveBaseEnv or something like that?

File: python/ray/rllib/agents/agent.py [hunk 1]
```diff
@@ -134,6 +134,9 @@
     # remote processes instead of in the same worker. This adds overheads, but
     # can make sense if your envs are very CPU intensive (e.g., for StarCraft).
     "remote_worker_envs": False,
+    # Similar to remote_worker_envs, but runs the envs asynchronously in the
+    # background for greater efficiency.
+    "async_remote_worker_envs": False,
```
Comment (richardliaw) [inline python/ray/rllib/agents/agent.py]: would be good to note that the two options are exclusive

File: python/ray/rllib/env/async_remote_env.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            actor = self.actors[env_id]
+            obj_id = actor.step.remote(actions)
+            self.pending[obj_id] = actor
+
+    def try_reset(self, env_id):
+        obs, _, _, _ = ray.get(self.actors[env_id].reset.remote())
+        return obs
+
+
+@ray.remote(num_cpus=0)
```
Comment (richardliaw) [inline python/ray/rllib/env/async_remote_env.py]: wouldn't be surprised if this requires GPU resources in the future (no need to address this rn)

**Review by richardliaw — APPROVED** [review]

File: python/ray/rllib/env/async_remote_env.py (hunk 2 above)
Comment (ericl) [reply python/ray/rllib/env/async_remote_env.py]: Yeah, I think so too.

**Pull request merged**